### PR TITLE
[SPARK-19774] StreamExecution should call stop() on sources when a stream fails

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

We call stop() on a Structured Streaming Source only when the stream is shutdown when a user calls streamingQuery.stop(). We should actually stop all sources when the stream fails as well, otherwise we may leak resources, e.g. connections to Kafka.

## How was this patch tested?

Unit tests in `StreamingQuerySuite`.

### CODE REVIEW

**Comment by brkyvz** [comment]
cc @tdas @zsxwing

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/util/MockSourceProvider.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+object MockSourceProvider {
+  // Function to generate sources. May provide multiple sources if the user implements such a
+  // function.
+  private var sourceProviderFunction: () => Source = _
+
+  final val fakeSchema = StructType(StructField("a", IntegerType) :: Nil)
+
+  def withMockSources(sources: Source*)(f: => Unit): Unit = {
+    require(sources.nonEmpty)
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/util/MockSourceProvider.scala]: nit: You can use `withMockSources(firstSource: Source, otherSources: Source*)` to make it become a compile error.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/util/MockSourceProvider.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    var i = 0
+    val srcProvider = () => {
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/util/MockSourceProvider.scala]: nit: `srcProvider` is not necessary. You can just assign the func to  `sourceProviderFunction`

**Review by zsxwing — COMMENTED** [review]
Looks good. Just some nits.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/util/MockSourceProvider.scala (hunk 1 above)
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/util/MockSourceProvider.scala]: good idea

**Comment by brkyvz** [comment]
@zsxwing Addressed

**Comment by zsxwing** [comment]
Thanks. LGTM. Merging to master and 2.1.

**Pull request closed without merging**